### DISCUSSION TITLE
some question about dfsu 3D

### DISCUSSION THREAD

**Discussion by lezhongwen** [DHI/mikeio#391] [Q&A]
Hello developers, this library has helped me solve a lot of problems recently, thank you for your contributions. I had some problems with reading and writing files.
1. The mesh data extracted from DFSU 3D files cannot be written to a mesh file, and i got ValueError: All elements must have 3 or 4 nodes. Element number 1 has 6 nodes.
2. After I wrote the regular mesh file(it can correspond to dfsu 2d) and the dataset extracted from DFSU 3D into a new DFSU file, the data became DFSU 2D. After checking, I found that not only the data quantity has decreased, but also the geometry attribute has changed.
I would be grateful for your reply~

**Comment by jsmariegaard** [comment]
Hi @lezhongwen - I am now back from vacation and will have a look at these issues. Thanks for reporting!

**Comment by jsmariegaard** [comment]
To write the mesh from a dfsu3d, the following can be used: 

     fn = "../tests/testdata/oresund_sigma_z.dfsu"
     dfs = mikeio.open(fn)
     dfs.to_mesh("oresund_sigma_z.mesh")

This should be equivalent to using `to_mesh()` method on the geometry, e.g. by 

     ds = mikeio.read(fn)
     ds.geometry.to_mesh("oresund_sigma_z.mesh")   # will fail

But this was not possible forr dfsu3d - I am fixing this in https://github.com/DHI/mikeio/pull/392.

**Reply by lezhongwen** [reply]
Thank you very much for your reply and contribution, and I am looking forward to the new version of Mikeio!!

**Comment by jsmariegaard** [comment]
@lezhongwen could you please elaborate on your 2nd point. I don't quite understand what you mean by "I found that not only the data quantity has decreased, but also the geometry attribute has changed" - could you please show an example?

**Reply by lezhongwen** [reply]
Dfsu3d should hold multiple layers of data in each grid of mesh. After I saved the data read from dfsu file with a mesh file(i am sure the mesh file can be matched with the data), I got a file with only one layer of data. And when the newly obtained file is read in, we can see its geometry attribute becomes dfSU2D.
